src/Rbac.php

```diff
@@ -134,27 +122,32 @@ public function getRole($objectOrName)
      * @param  RoleInterface|string             $role
      * @param  string                           $permission
      * @param  AssertionInterface|Callable|null $assert
-     * @throws Exception\InvalidArgumentException
      * @return bool
+     * @throws Exception\InvalidArgumentException
      */
-    public function isGranted($role, $permission, $assert = null)
+    public function isGranted($role, string $permission, $assert = null): bool
     {
-        $result = $this->getRole($role)->hasPermission($permission);
-
-        if ($assert) {
-            if ($assert instanceof AssertionInterface) {
-                return $result && $assert->assert($this);
-            }
-
-            if (is_callable($assert)) {
-                return $result && $assert($this);
-            }
-
-            throw new Exception\InvalidArgumentException(
-                'Assertions must be a Callable or an instance of Zend\Permissions\Rbac\AssertionInterface'
-            );
+        if (! $this->hasRole($role)) {
+            throw new Exception\InvalidArgumentException(sprintf(
+                'No role with name "%s" could be found',
+                is_object($role) ? $role->getName() : $role
+            ));
         }
-
-        return $result;
+        if (is_string($role)) {
+            $role = $this->getRole($role);
+        }
+        $result = $role->hasPermission($permission);
+        if (false === $result || null === $assert) {
+            return $result;
+        }
+        if ($assert instanceof AssertionInterface) {
+            return $result && $assert->assert($this, $permission, $role);
+        }
+        if (is_callable($assert)) {
+            return $result && $assert($this, $permission, $role);
+        }
```

Comment:
Considering there's already the `CallbackAssertion`, it might be easier to declare:

```php
AssertionInterface $assert = null
```

and require developers to decorate callable assertions before passing them to this method.

Reply:
I think we can limit the BC breaks here if we assume that `$assert` can be `AssertionInterface` or `Callable` or `null`, like the actual implementation.